src/sync/ms_queue.rs

```diff
@@ -34,49 +68,209 @@ impl<T> MsQueue<T> {
-                    n = owned;
+            }
+        }
+    }
+
+    #[inline(always)]
+    // Attempt to pop a data node. `Ok(None)` if queue is empty or in blocking
+    // mode; `Err(())` if lost race to pop.
+    fn pop_internal(&self, guard: &epoch::Guard) -> Result<Option<T>, ()> {
+        let head = self.head.load(Acquire, guard).unwrap();
+        if let Some(next) = head.next.load(Acquire, guard) {
+            if let Payload::Data(ref t) = next.payload {
+                unsafe {
+                    if self.head.cas_shared(Some(head), Some(next), Release) {
+                        guard.unlinked(head);
+                        Ok(Some(ptr::read(t)))
+                    } else {
+                        Err(())
+                    }
                 }
+            } else {
+                Ok(None)
             }
+        } else {
+            Ok(None)
         }
     }
 
     /// Attempt to dequeue from the front.
     ///
     /// Returns `None` if the queue is observed to be empty.
-    pub fn pop(&self) -> Option<T> {
+    pub fn try_pop(&self) -> Option<T> {
         let guard = epoch::pin();
         loop {
-            let head = self.head.load(Acquire, &guard).unwrap();
+            if let Ok(r) = self.pop_internal(&guard) {
+                return r;
+            }
+        }
+    }
 
-            if let Some(next) = head.next.load(Acquire, &guard) {
-                unsafe {
-                    if self.head.cas_shared(Some(head), Some(next), Release) {
-                        guard.unlinked(head);
-                        return Some(ptr::read(&(*next).data))
+    /// Dequeue an element from the front of the queue, blocking if the queue is
+    /// empty.
+    pub fn pop(&self) -> T {
+        let guard = epoch::pin();
+
+        // Fast path: keep retrying until we observe that the queue has no data,
+        // avoiding the allocation of a blocked node.
+        loop {
+            match self.pop_internal(&guard) {
+                Ok(Some(r)) => {
+                    return r;
+                }
+                Ok(None) => {
+                    break;
+                }
+                Err(()) => {}
+            }
+        }
+
+        // The signal gets to live on the stack, since this stack frame will be
+        // blocked until receiving the signal.
+        let mut signal = Signal {
+            thread: thread::current(),
+            data: None,
+            ready: AtomicBool::new(false),
+        };
+
+        // Go ahead and allocate the blocked node; chances are, we'll need it.
+        let mut node = Owned::new(Node {
+            payload: Payload::Blocked(&mut signal),
+            next: Atomic::null(),
+        });
+
+        loop {
+            // try a normal pop
+            if let Ok(Some(r)) = self.pop_internal(&guard) {
+                return r;
+            }
+
+            let head = self.head.load(Relaxed, &guard).unwrap();
+            let tail = self.tail.load(Relaxed, &guard).unwrap();
```

Comment:
I remember when we were talking about this we concluded that it was important the load of tail happened before the load of the head. I can't quite recall why, but was the swap here intentional?